GuitarLSTM/LSTM.ipynb

```diff
@@ -1,67 +1,235 @@
+    "\n",
+    "    # needed for stateful LSTM\n",
+    "    self.hidden = None  # (hidden state, cell state)\n",
+    "\n",
+    "  def forward(self, x):\n",
+    "    \"\"\"\n",
+    "    :param x: tensor of shape (batch_size, seq_length, features)\n",
+    "    \"\"\"\n",
+    "    if self.hidden is None:\n",
+    "      self.hidden = (\n",
+    "        torch.zeros(self.num_layers, x.size(0), self.hidden_size).to(x.device),\n",
+    "        torch.zeros(self.num_layers, x.size(0), self.hidden_size).to(x.device)\n",
+    "      )\n",
+    "    \n",
+    "    out, self.hidden = self.lstm(x, self.hidden)\n",
+    "\n",
+    "    out = self.fc(out)\n",
+    "\n",
+    "    # add residual step\n",
+    "    if self.skip:\n",
+    "      out += x[:, :, :]\n",
+    "    \n",
+    "    return out\n",
+    "  \n",
+    "  def reset_hidden(self):\n",
+    "    self.hidden = None\n",
+    "\n",
+    "  # detach hidden state, this resets gradient tracking on the hidden state\n",
+    "  def detach_hidden(self):\n",
+    "    if self.hidden.__class__ == tuple:\n",
+    "      self.hidden = tuple([h.clone().detach() for h in self.hidden])\n",
+    "    else:\n",
+    "      self.hidden = self.hidden.clone().detach()\n",
+    "\n",
+    "  def train_epoch(self, input_data, target_data, loss_fcn, optim, bs, init_len=200, up_fr=1000):\n",
+    "    \"\"\"\n",
+    "    :param input_data: x examples with shape (num_examples, seq_length, features)\n",
+    "    :param target_data: y target with shape (num_examples, seq_length, features)\n",
+    "    :param loss_fcn: our loss function\n",
+    "    :param optim: our optimizer\n",
+    "    :param bs: batch size, number of examples per batch\n",
+    "    :param init_len: number of audio samples to initialize hidden state\n",
+    "    :param up_fr: number of timesteps before we backprop (truncated backprop)\n",
+    "    \"\"\"\n",
+    "    # shuffle the examples around\n",
+    "    shuffle = torch.randperm(input_data.shape[0])\n",
+    "\n",
+    "    # Iterate over the batches\n",
+    "    ep_loss = 0\n",
+    "    for batch_i in range(math.ceil(shuffle.shape[0] / bs)):\n",
+    "      # Load batch of shuffled segments\n",
+    "      input_batch = input_data[shuffle[batch_i * bs:(batch_i + 1) * bs], :, :]\n",
+    "      target_batch = target_data[shuffle[batch_i * bs:(batch_i + 1) * bs],:, :]\n",
+    "\n",
+    "      # Initialise network hidden state by processing some samples then zero the gradient buffers\n",
+    "      self(input_batch[:, 0:init_len, :])\n",
+    "      self.zero_grad()\n",
+    "\n",
+    "      # Choose the starting index for processing the rest of the batch sequence, in chunks of args.up_fr\n",
+    "      start_i = init_len\n",
+    "      batch_loss = 0\n",
+    "      k = 0\n",
+    "      # Iterate over the remaining samples in the mini batch\n",
+    "      for k in range(math.ceil((input_batch.shape[0] - init_len) / up_fr)):\n",
```

Comment:
the reason why the loss is 0 is because math.ceil((input_batch.shape[0] - init_len) / up_fr) is 0 because  batch_size - inin_len<0 when batch_size=10 and init_len=200

Reply:
Oh I see the issue, this should be "input_batch.shape[1]". Nice catch!